### PR TITLE
Add an __VERIFIER_expected_fail() intrinsic, and use it in a test

### LINKED ISSUE #228: It should be possible to have assertions which are expected to fail in RMC tests
Requested feature: What it says in the title
Use case: Makes it easier to write tests, gets rid of need for `Expected` files
Link to relevant documentation (Rust reference, Nomicon, RFC):
Is this a breaking change? No

Test case:

```rust
let x = __nondet();
__VERIFIER_assume(x < 10);
expect_fail(x == 12); // Should lead to a failing assertion, which should lead to the test passing
```

### PR DESCRIPTION
**Pull request by danielsn**
### Description of changes: 

Currently, we check for expected failing assertions by marking the whole test as `_fail`. This is not very granular.
### Resolved issues:

Resolves #228 

### Call-outs:

We have a field for custom error messages, but due to the inability of CBMC to take non-constant assertion messages, we can't yet use them.

### Testing:

* How is this change tested? Used it in a test, test still fails

* Is this a refactor change?

### Checklist
- [ ] Each commit message has a non-empty body, explaining why the change was made
- [ ] Methods or procedures are documented
- [ ] Regression or unit tests are included, or existing tests cover the modified code
- [ ] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

### CODE REVIEW

File: compiler/rustc_codegen_llvm/src/gotoc/hooks.rs [hunk 1]
```diff
@@ -74,6 +74,46 @@ fn name_is(tcx: TyCtxt<'tcx>, instance: Instance<'tcx>, expected: &str) -> bool
...
+    ) -> Stmt {
+        assert_eq!(fargs.len(), 2);
+        let target = target.unwrap();
+        let cond = fargs.remove(0).cast_to(Type::bool());
+        //TODO: actually use the error message passed by the user.
+        let msg = "EXPECTED FAIL";
+        let loc = tcx.codegen_span_option2(span);
+        Stmt::block(
+            vec![
+                Stmt::assert(cond, msg, loc.clone()),
```
Comment (adpaco-aws) [inline compiler/rustc_codegen_llvm/src/gotoc/hooks.rs:109]: I do not understand why we are unable to use the error message here. You mention CBMC is not capable of taking non-constant messages, but the one included below is constant, right?

File: src/test/cbmc/Assume/main_fail.rs [hunk 2]
```diff
@@ -9,8 +9,12 @@ fn __nondet<T>() -> T {
     unimplemented!()
 }
 
+fn __VERIFIER_expect_fail(cond: bool, msg: &str)  {
+    unimplemented!()
+}
+
```
Comment (adpaco-aws) [inline src/test/cbmc/Assume/main_fail.rs:15]: This should be included in the new RMC prelude.

**Review by adpaco-aws — CHANGES_REQUESTED** [review]

File: compiler/rustc_codegen_llvm/src/gotoc/hooks.rs (hunk 1 above)
Comment (danielsn) [reply compiler/rustc_codegen_llvm/src/gotoc/hooks.rs:109]: The value is a constant. But Rust puts it inside a variable, and then passes the address of the variable. So we need to figure out how to reverse that.

File: src/test/cbmc/Assume/main_fail.rs (hunk 2 above)
Comment (danielsn) [reply src/test/cbmc/Assume/main_fail.rs:15]: Willdo

File: compiler/rustc_codegen_llvm/src/gotoc/hooks.rs (hunk 1 above)
Comment (danielsn) [reply compiler/rustc_codegen_llvm/src/gotoc/hooks.rs:109]: https://github.com/model-checking/rmc/issues/236

**Review by adpaco-aws — APPROVED** [review]

**Pull request merged**